src/apps/monero/protocol/tsx_sign_builder.py

```diff
@@ -0,0 +1,1416 @@
+        tmp_buff = bytearray(32)
+        rsig_data.mask = bytearray(32 * self.num_dests())
+        self.sumout = crypto.sc_init(0)
+        for i in range(self.num_dests()):
+            cur_mask = crypto.new_scalar()
+            is_last = i + 1 == self.num_dests()
+            if is_last and self.use_simple_rct:
+                crypto.sc_sub_into(cur_mask, self.sumpouts_alphas, self.sumout)
+            else:
+                crypto.random_scalar(cur_mask)
+
+            crypto.sc_add_into(self.sumout, self.sumout, cur_mask)
+            self.output_masks.append(cur_mask)
+            crypto.encodeint_into(tmp_buff, cur_mask)
+            utils.memcpy(rsig_data.mask, 32 * i, tmp_buff, 0, 32)
+
+        self.assrt(crypto.sc_eq(self.sumout, self.sumpouts_alphas), "Invalid masks sum")
+        self.sumout = crypto.sc_init(0)
+        return resp
+
+    def _get_out_mask(self, idx):
+        if self.rsig_offload:
+            return self.output_masks[idx]
+        else:
+            is_last = idx + 1 == self.num_dests()
+            if is_last:
+                return crypto.sc_sub(self.sumpouts_alphas, self.sumout)
+            else:
+                return crypto.random_scalar()
+
+    def _get_rsig_batch(self, idx):
+        """
+        Returns index of the current rsig batch
+        """
+        r = 0
+        c = 0
+        while c < idx + 1:
+            c += self.rsig_grp[r]
+            r += 1
+        return r - 1
+
+    def _is_last_in_batch(self, idx, bidx=None):
+        """
+        Returns true if the current output is last in the rsig batch
+        """
+        bidx = self._get_rsig_batch(idx) if bidx is None else bidx
+        batch_size = self.rsig_grp[bidx]
+        return (idx - sum(self.rsig_grp[:bidx])) + 1 == batch_size
+
+    def _gen_commitment(self, in_amount):
+        """
+        Computes Pedersen commitment - pseudo outs
+        Here is slight deviation from the original protocol.
+        We want that \\sum Alpha = \\sum A_{i,j} where A_{i,j} is a mask from range proof for output i, bit j.
+
+        Previously this was computed in such a way that Alpha_{last} = \\sum A{i,j} - \\sum_{i=0}^{last-1} Alpha
+        But we would prefer to compute commitment before range proofs so alphas are generated completely randomly
+        and the last A mask is computed in this special way.
+        Returns pseudo_out
+        """
+        alpha = crypto.random_scalar()
+        self.sumpouts_alphas = crypto.sc_add(self.sumpouts_alphas, alpha)
+        return alpha, crypto.gen_commitment(alpha, in_amount)
+
+    def _check_out_commitment(self, amount, mask, C):
+        self.assrt(
+            crypto.point_eq(
+                C,
+                crypto.point_add(
+                    crypto.scalarmult_base(mask), crypto.scalarmult_h(amount)
+                ),
+            ),
+            "OutC fail",
+        )
+
+    def _check_bproof(self, batch_size, rsig, masks):
```

Comment:
I think this function is not being used?

Reply:
Correct, it can be removed, the offloaded bulletproof is sent without commitment values so there is no check needed.